olm/operator-registry.md

```diff
@@ -0,0 +1,324 @@
+
+Internal details:
+Polling process should be reviewed by quay team to ensure we scale on quay (although these indexes should work on ANY container image registry)
+
+### Implementation Details
+
+The goal of this enhancement is to replace the app-registry implementation in a way that is usable by operator developers and CI pipeline developers. To do this we will build an index image that is defined from a set of operator bundle images that are released independently and optimized in a way that can be productized by operator authors and pipeline developers.
+
+This implementation makes the assumption that a separate process is built that generates container images that contain individual operator bundles for each version of an operator. The intent of the index image is to aggregate these images that are out of scope of this enhancement. Once that image is built, the image itself is immutable. In order to create a new image, the tooling described here will run off cluster. Once the image is loaded onto the cluster, the only way to load new content is to pull down a new version of that image (either manually or by an automated poll).
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
```

Comment:
What's the behavior when an existing, indexed bundle is getting updated and supplied to the `add` command - is there some sort of `INSERT ON DUPLICATE KEY UPDATE` behavior?
Background: it's going to be fairly hard for a pipeline to determine whether the bundle that just got pushed is a net-net version or replacing an existing version.

Reply:
The behavior is that it will fail. This is the same existing enforcement that currently exists about not being able to easily update an existing version -- once that version is released into the wild modifying it is problematic because what happens when someone has already installed that version? Instead if another change is required just bump the version again.

If there is some more serious reason why that version needs to be removed entirely, then the safety valve is a manual process of deleting that image version and either modifying the index by deleting the version and then adding the new one OR deleting the image and rebuilding from scratch.

Reply:
Thanks. That also means, that any cosmetic changes, e.g. typos in descriptions, or logos, will lead to an updated bundle which leads to an updated version which leads to updates happening on cluster. Unless we modify / recreate the index manually as you describe.
Just making sure we acknowledge this. From my side this is fine.